### PR TITLE
Integrate Flex Decoding

### PR DESCRIPTION
**Pull request by BoyuanFeng**
This PR integrates flex decoding with gpt-fast.

### End-to-end performance gain of Llama2-7b

Device: H100
Unit: tokens/sec

Length | spda | Flex Decoding | Speedup
-- | -- | -- | --
1024 | 143.57 | 142.59 | 0.99x
2048 | 138.67 | 140.13 | 1.01x
4096 | 128.85 | 135.31 | 1.05x
8192 | 111.59 | 125.38 | 1.12x
16384 | 89 | 109.8 | 1.23x
32768 | 64.11 | 87.67 | 1.37x

command:
```
export MODEL_REPO=meta-llama/Llama-2-7b-chat-hf
python generate.py --compile --max_new_tokens 16384 --checkpoint_path checkpoints/$MODEL_REPO/model.pth --prompt "Hello, my name is"
```

Please also set `ModelArgs.block_size = 65536` to repeat the result.

We expect to see larger speedup on longer context length.

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -120,11 +159,9 @@ def setup_caches(self, max_batch_size, max_seq_length):
             b.attention.kv_cache = KVCache(max_batch_size, max_seq_length, self.config.n_local_heads, head_dim, dtype)
 
         self.freqs_cis = precompute_freqs_cis(self.config.block_size, self.config.dim // self.config.n_head, self.config.rope_base, dtype)
-        self.causal_mask = torch.tril(torch.ones(self.max_seq_length, self.max_seq_length, dtype=torch.bool))
 
-    def forward(self, idx: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
+    def _forward(self, mask: BlockMask, idx: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
```
Comment (Chillee) [inline model.py:163]: Why was this refactor done?

File: model.py [hunk 2]
```diff
@@ -10,13 +10,51 @@
...
+    def __init__(self, max_seq_length):
+        self.input_pos = None
+        self.max_seq_length = max_seq_length
+        self.block_masks = create_block_mask(causal_mask, 1, 1, max_seq_length, max_seq_length, device="cuda")
+
+    def decoding_causal_mask(self, b, h, q, kv):
+        offset = self.input_pos[0]
+        return offset + q >= kv
+
+    def get_mask(self, kv_len, input_pos) -> BlockMask:
```
Comment (Chillee) [inline model.py:36]: I think we should factor out indexing and support it on BlockMask directly.

File: model.py (hunk 1 above)
Comment (Chillee) [reply model.py:134]: > For prefill, we have various length of BlockMask, leading to a dynamic length slicing issue with dynamo.
Why is this a problem? Can't we just index into BlockMask with `input_pos`?

File: model.py [hunk 3]
```diff
@@ -89,7 +103,7 @@ def update(self, input_pos, k_val, v_val):
         return k_out, v_out
 
 class Transformer(nn.Module):
-    def __init__(self, config: ModelArgs) -> None:
+    def __init__(self, config: ModelArgs, get_mask_mod: Callable[[int], _mask_mod_signature]) -> None:
```
Comment (Chillee) [inline model.py:106]: `get_mask_mod` shouldn't take an integer - it should take a `mask_mod`. We also don't need to set at as an argument, just set it as an attribute within the module.

Comment (Chillee) [reply model.py:106]: Specifically, you should be able to take any existing `mask_mod` and wrap it to make it automatically support an offset.

File: generate.py [hunk 4]
```diff
@@ -54,29 +55,35 @@ def sample(logits, temperature: float = 1.0, top_k: Optional[int] = None):
...
+    logits = model(mask, x, input_pos)
     return sample(logits, **sampling_kwargs)[0]
 
 def decode_one_token(model: Transformer, x: torch.Tensor, input_pos: torch.Tensor, **sampling_kwargs) -> Tuple[torch.Tensor, torch.Tensor]:
     # input_pos: [B, 1]
     assert input_pos.shape[-1] == 1
-    logits = model(x, input_pos)
+    block_index = input_pos // model.block_mask.BLOCK_SIZE[0]
+    mask = model.block_mask[block_index]
+    mask.mask_mod = model.get_mask_mod(input_pos[0])
```
Comment (Chillee) [inline generate.py:72]: I would just automatically transform it here to support an offset.

**Review by Chillee — APPROVED** [review]

**Review by Chillee — CHANGES_REQUESTED** [review]
Mostly looks good (modulo some other nits), although we'll probably want to wait on landing this until the other PRs on pytorch core are landed at least.

**Comment by merrymercy** [comment]
This looks interesting. I would like to share some numbers we got with torch.compile + flashinfer in sglang. It can serve as some good baselines.

```
# length = 1024
# Decode.  latency: 0.00594 s, throughput:    168.31 token/s
python3 -m sglang.bench_latency --model meta-llama/Llama-2-7b-chat-hf --batch-size 1 --input 1024 --output 8 --enable-torch-compile

# length = 32768
# Decode.  latency: 0.01133 s, throughput:     88.27 token/s
python3 -m sglang.bench_latency --model meta-llama/Llama-2-7b-chat-hf --batch-size 1 --input 1024 --output 8 --enable-torch-compile
```

You can find more numbers at https://github.com/sgl-project/sglang/issues/1008

**Comment by Chillee** [comment]
@merrymercy We run on nerfed H100s internally at Meta with only 2.4 TB/s of bandwidth, so these numbers aren't 1:1 comparable.

File: generate.py [hunk 5]
```diff
@@ -154,10 +164,11 @@ def generate(
     # create an empty tensor of the expected final shape and fill in the current tokens
     T = prompt.size(0)
     T_new = T + max_new_tokens
+    T_buf = roundup(T_new, 128) # round up to multiple of 128 to use flex_attention
```
Comment (Chillee) [inline generate.py:167]: Is this still needed?

File: generate.py [hunk 6]
```diff
@@ -54,29 +55,38 @@ def sample(logits, temperature: float = 1.0, top_k: Optional[int] = None):
...
+    mask = create_block_mask(causal_mask, 1, 1, input_pos.shape[0], model.max_seq_length, device="cuda")
+    logits = model(mask, x, input_pos)
     return sample(logits, **sampling_kwargs)[0]
 
-def decode_one_token(model: Transformer, x: torch.Tensor, input_pos: torch.Tensor, **sampling_kwargs) -> Tuple[torch.Tensor, torch.Tensor]:
+def decode_one_token(model: Transformer, x: torch.Tensor, input_pos: torch.Tensor, block_mask: BlockMask, **sampling_kwargs) -> Tuple[torch.Tensor, torch.Tensor]:
     # input_pos: [B, 1]
     assert input_pos.shape[-1] == 1
-    logits = model(x, input_pos)
+    block_index = input_pos // block_mask.BLOCK_SIZE[0]
```
Comment (Chillee) [inline generate.py:73]: Shouldn't we do this divide inside the block mask indexing?

File: generate.py [hunk 7]
```diff
@@ -56,29 +57,39 @@ def sample(logits, temperature: float = 1.0, top_k: Optional[int] = None):
...
+    mask = block_mask[:, :, block_index]
+    mask.mask_mod = block_mask.mask_mod
+    logits = model(mask, x, input_pos)
     return sample(logits, **sampling_kwargs)
 
 def decode_n_tokens(model: Transformer, cur_token: torch.Tensor, input_pos: torch.Tensor, num_new_tokens: int, callback=lambda _: _, **sampling_kwargs):
+    block_mask = create_block_mask(causal_mask, 1, 1, model.max_seq_length, model.max_seq_length, device="cuda")
```
Comment (drisspg) [inline generate.py:82]: try doing 
create_block_mask_compile = torch.compile(create_block_mask)

as a global
